### DISCUSSION TITLE
1614 with DMXserial library

### DISCUSSION THREAD

**Discussion by CircuitSmuggler** [SpenceKonde/megaTinyCore#462] [Library discussion]
I've got a attiny1614 and would like to use the DMXserial library.  Quickly found out from all the compiling errors that it doesn't work. Would anyone know if this is an easy port over or would it take a skilled hand? BTW I'm somewhat of a novice so unless its just changing a few things in the .h file its probably far beyond my skill level.  Here are the error messages I'm getting.

`Arduino: 1.8.13 (Windows 10), Board: "ATtiny1614/1604/814/804/414/404 w/Optiboot, ATtiny1614, 20 MHz internal, Enabled (default timer), 1.8V (5 MHz or less), Disabled/Disabled, Ignore (saves flash, almost always fine), TX:5 (PB2), RX:4 (PB3), UPDI, Optiboot for 8 sec. after power-on & reset, 8ms"

In file included from C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial.cpp:69:0:

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h: In function 'void _DMX_init()':

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:47:11: error: 'PORTMUX_t {aka struct PORTMUX_struct}' has no member named 'USARTROUTEA'

   PORTMUX.USARTROUTEA |= PORTMUX_USART1_ALT1_gc;

           ^~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:47:26: error: 'PORTMUX_USART1_ALT1_gc' was not declared in this scope

   PORTMUX.USARTROUTEA |= PORTMUX_USART1_ALT1_gc;

                          ^~~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:47:26: note: suggested alternative: 'PORTMUX_USART0_DEFAULT_gc'

   PORTMUX.USARTROUTEA |= PORTMUX_USART1_ALT1_gc;

                          ^~~~~~~~~~~~~~~~~~~~~~

                          PORTMUX_USART0_DEFAULT_gc

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:50:4: error: 'USART1' was not declared in this scope

   (USART1).CTRLB = USART_RXMODE_NORMAL_gc;

    ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:50:4: note: suggested alternative: 'USART0'

   (USART1).CTRLB = USART_RXMODE_NORMAL_gc;

    ^~~~~~

    USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:53:11: error: 'PIN_WIRE_HWSERIAL1_RX' was not declared in this scope

   pinMode(PIN_WIRE_HWSERIAL1_RX, INPUT_PULLUP);

           ^~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:53:11: note: suggested alternative: 'PIN_HWSERIAL0_RX'

   pinMode(PIN_WIRE_HWSERIAL1_RX, INPUT_PULLUP);

           ^~~~~~~~~~~~~~~~~~~~~

           PIN_HWSERIAL0_RX

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:54:11: error: 'PIN_WIRE_HWSERIAL1_TX' was not declared in this scope

   pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT);

           ^~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:54:11: note: suggested alternative: 'PIN_HWSERIAL0_TX'

   pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT);

           ^~~~~~~~~~~~~~~~~~~~~

           PIN_HWSERIAL0_TX

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h: In function 'void _DMX_setMode(DMXUARTMode)':

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:74:6: error: 'USART1' was not declared in this scope

     (USART1).CTRLB = USART_RXMODE_NORMAL_gc; // (USART_RXEN_bm | USART_TXEN_bm);

      ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:74:6: note: suggested alternative: 'USART0'

     (USART1).CTRLB = USART_RXMODE_NORMAL_gc; // (USART_RXEN_bm | USART_TXEN_bm);

      ^~~~~~

      USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:78:6: error: 'USART1' was not declared in this scope

     (USART1).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

      ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:78:6: note: suggested alternative: 'USART0'

     (USART1).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

      ^~~~~~

      USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:84:6: error: 'USART1' was not declared in this scope

     (USART1).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

      ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:84:6: note: suggested alternative: 'USART0'

     (USART1).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

      ^~~~~~

      USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:91:6: error: 'USART1' was not declared in this scope

     (USART1).CTRLB = 0; // no operation

      ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:91:6: note: suggested alternative: 'USART0'

     (USART1).CTRLB = 0; // no operation

      ^~~~~~

      USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:97:13: error: 'PIN_WIRE_HWSERIAL1_TX' was not declared in this scope

     pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT); // is required again after disabling UART

             ^~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:97:13: note: suggested alternative: 'PIN_HWSERIAL0_TX'

     pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT); // is required again after disabling UART

             ^~~~~~~~~~~~~~~~~~~~~

             PIN_HWSERIAL0_TX

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:103:6: error: 'USART1' was not declared in this scope

     (USART1).CTRLA = 0; // disable all interrupts

      ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:103:6: note: suggested alternative: 'USART0'

     (USART1).CTRLA = 0; // disable all interrupts

      ^~~~~~

      USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:109:6: error: 'USART1' was not declared in this scope

     (USART1).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

      ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:109:6: note: suggested alternative: 'USART0'

     (USART1).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

      ^~~~~~

      USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h: In function 'void _DMX_flush()':

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:123:23: error: 'USART1' was not declared in this scope

   uint8_t voiddata = (USART1).RXDATAL;

                       ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:123:23: note: suggested alternative: 'USART0'

   uint8_t voiddata = (USART1).RXDATAL;

                       ^~~~~~

                       USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h: In function 'void _DMX_writeByte(uint8_t)':

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:131:4: error: 'USART1' was not declared in this scope

   (USART1).TXDATAL = data;

    ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:131:4: note: suggested alternative: 'USART0'

   (USART1).TXDATAL = data;

    ^~~~~~

    USART0

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h: In function 'void USART1_RXC_vect()':

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:140:25: error: 'USART1' was not declared in this scope

   register8_t rxferr = (USART1).RXDATAH & USART_FERR_bm;

                         ^~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:140:25: note: suggested alternative: 'USART0'

   register8_t rxferr = (USART1).RXDATAH & USART_FERR_bm;

                         ^~~~~~

                         USART0
`

**Comment by SpenceKonde** [comment]
Excellent news: That library already supports the modern AVR architecture. It just makes some bad assumptions (mainly that there's more than 1 serial port, I'll bet it would come a lot closert to compiling - or might even build straight up - for the 1624 just because of the second serial port. 

I forked it and took a quick stab at it. 
https://github.com/SpenceKonde/DMXSerial/

For parts other than the 4809, you'd wantto examine DMXSerial_megaavr.h and adjust it to match your needs.

**Reply by SpenceKonde** [reply]
haven't done any sort of test or anything

**Reply by SpenceKonde** [reply]
My hope is that we can PR it and fix it for everyone

**Comment by CircuitSmuggler** [comment]
Thanks for looking into it! I'll see if can have time to test it today.

**Comment by CircuitSmuggler** [comment]
So I got some 1624 just to give it a try.  I'm getting new  but much less compiling errors. Seems a little more promising.   Hoping these are easy fixes.  

`Arduino: 1.8.13 (Windows 10), Board: "ATtiny3224/1624/1614/1604/824/814/804/424/414/404/214/204, ATtiny1624, 20 MHz internal, 1.8V (5 MHz or less), Disabled/Disabled, EEPROM retained, TCB1 (1614, 2-series only), UPDI (no reset pin), 8ms, Ignore (saves flash, almost always fine)"

In file included from C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial.cpp:69:0:

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h: In function 'void _DMX_init()':

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:32:26: error: 'SIGROW_t {aka struct SIGROW_struct}' has no member named 'OSC16ERR5V'

   int8_t oscErr = SIGROW.OSC16ERR5V;

                          ^~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:47:26: error: 'PORTMUX_USART1_ALT1_gc' was not declared in this scope

   PORTMUX.USARTROUTEA |= PORTMUX_USART1_ALT1_gc;

                          ^~~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:47:26: note: suggested alternative: 'PORTMUX_USART0_ALT1_gc'

   PORTMUX.USARTROUTEA |= PORTMUX_USART1_ALT1_gc;

                          ^~~~~~~~~~~~~~~~~~~~~~

                          PORTMUX_USART0_ALT1_gc

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:53:11: error: 'PIN_WIRE_HWSERIAL1_RX' was not declared in this scope

   pinMode(PIN_WIRE_HWSERIAL1_RX, INPUT_PULLUP);

           ^~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:53:11: note: suggested alternative: 'PIN_HWSERIAL1_RX'

   pinMode(PIN_WIRE_HWSERIAL1_RX, INPUT_PULLUP);

           ^~~~~~~~~~~~~~~~~~~~~

           PIN_HWSERIAL1_RX

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:54:11: error: 'PIN_WIRE_HWSERIAL1_TX' was not declared in this scope

   pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT);

           ^~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:54:11: note: suggested alternative: 'PIN_HWSERIAL1_TX'

   pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT);

           ^~~~~~~~~~~~~~~~~~~~~

           PIN_HWSERIAL1_TX

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h: In function 'void _DMX_setMode(DMXUARTMode)':

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:97:13: error: 'PIN_WIRE_HWSERIAL1_TX' was not declared in this scope

     pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT); // is required again after disabling UART

             ^~~~~~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerial\src\DMXSerial_megaavr.h:97:13: note: suggested alternative: 'PIN_HWSERIAL1_TX'

     pinMode(PIN_WIRE_HWSERIAL1_TX, OUTPUT); // is required again after disabling UART

             ^~~~~~~~~~~~~~~~~~~~~

             PIN_HWSERIAL1_TX

exit status 1

Error compiling for board ATtiny3224/1624/1614/1604/824/814/804/424/414/404/214/204.



This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.
`

**Reply by SpenceKonde** [reply]
I already posted a version that I expect will work on both the 1614 and 1624, as well as (after entering some simple chip/application specific parameters in the megaavr header file, which is always going to be needed one way or another, and there isn't really any way to get preprocessor directives into a library in Arduino IDE which is what you'd need. ) every AVR device released since 2016 (the ones with the new architecture). It fully takes over whichever UART you tell it to use - if it's using UARTNUM is 0, Serial is unavailable, if UARTNUM is 1 Serial1 is unavailable, etc and any even trivial references to it will cause it to try to include the ISRs from the file and give duplicate vector errors. 

But those error messages you are posting are not coming from my version of the library, because I changed a bunch of that code. Please try the version of that library I posted. Three days ago. 

If you want Serial and DMX at the same time, you do definitely need something with two serial ports

**Comment by CircuitSmuggler** [comment]
Thanks for clearing that up for me.  I set up your version and altered the parameters for the chip I'm using.  Still getting errors messages.  I was able to solve the simple one of a definition missing.  But the other ones have me kind of stumped, especially the SIGROW issue. Really, thank you for taking the time to help out. 

`Arduino: 1.8.15 (Windows 10), Board: "ATtiny3224/1624/1614/1604/824/814/804/424/414/404/214/204, ATtiny1624, 16 MHz internal, 1.8V (5 MHz or less), Disabled/Disabled, EEPROM retained, Enabled (default timer), UPDI (no reset pin), 8ms, Ignore (saves flash, almost always fine)"


In file included from C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial.cpp:69:0:

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h: In function 'void _DMX_init()':

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:85:30: error: 'SIGROW_t {aka struct SIGROW_struct}' has no member named 'OSC16ERR5V'

       int8_t oscErr = SIGROW.OSC16ERR5V;

                              ^~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:91:5: error: '_DMX_dmxDivider' was not declared in this scope

     _DMX_dmxDivider = (64 * F_CPU) / (16 * baud);

     ^~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:95:5: error: '_DMX_breakDivider' was not declared in this scope

     _DMX_breakDivider = (64 * F_CPU) / (16 * baud);

     ^~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:33:26: error: 'PORTMUX_USART1_ALT1_gc' was not declared in this scope

 #define _DMX_USARTMUXOPT PORTMUX_USART1_ALT1_gc

                          ^

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:112:28: note: in expansion of macro '_DMX_USARTMUXOPT'

     PORTMUX.USARTROUTEA |= _DMX_USARTMUXOPT;

                            ^~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:33:26: note: suggested alternative: 'PORTMUX_USART0_ALT1_gc'

 #define _DMX_USARTMUXOPT PORTMUX_USART1_ALT1_gc

                          ^

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:112:28: note: in expansion of macro '_DMX_USARTMUXOPT'

     PORTMUX.USARTROUTEA |= _DMX_USARTMUXOPT;

                            ^~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h: In function 'void _DMX_setMode(DMXUARTMode)':

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:147:34: error: '_DMX_dmxDivider' was not declared in this scope

     (_DMX_USART).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

                                  ^~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:153:34: error: '_DMX_dmxDivider' was not declared in this scope

     (_DMX_USART).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

                                  ^~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:163:34: error: '_DMX_breakDivider' was not declared in this scope

     (_DMX_USART).BAUD = (int16_t)_DMX_breakDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

                                  ^~~~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:173:34: error: '_DMX_dmxDivider' was not declared in this scope

     (_DMX_USART).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

                                  ^~~~~~~~~~~~~~~

C:\Users\\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:178:34: error: '_DMX_dmxDivider' was not declared in this scope

     (_DMX_USART).BAUD = (int16_t)_DMX_dmxDivider; // assign the baud_divider, a.k.a. BAUD (USART Baud Rate Register)

                                  ^~~~~~~~~~~~~~~

exit status 1

Error compiling for board ATtiny3224/1624/1614/1604/824/814/804/424/414/404/214/204.



This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.
`

**Reply by SpenceKonde** [reply]
#define _DMX_USARTMUXOPT PORTMUX_USART1_ALT1_gc

you must choose a mux option that exists; on the 1624, you don't have an ALT1 portmux setting - the alt pins are on PC1 and PC2. Replace that with PORTMUX_USART1_DEFAULT_gc (which so happens to be 0; all the default PORTMUX group codes are. 

```
C:\Users\Documents\Arduino\libraries\DMXSerialTINYmaster\src\DMXSerial_megaavr.h:85:30: error: 'SIGROW_t {aka struct SIGROW_struct}' has no member named 'OSC16ERR5V'
```
is a bug in the version of the ATPACKs from Microchip (theyh were missing a bunch of members of the SIGROW struct in the headers (hardware akways had them)What's weird IMO is that my code isn't detecting and  working around that. I sm doiing a releqase soon to pick up a fix for issues with Serial UPDI not working at higher baud rates tuned internal oscillator support.... tthat version will incorporate the latest toolchain version. Plus some stupid SPI bug. 

You can replace the OSC20ERR...  or OPSC15ERR... with 0 on the 2-series and feel pretty safe doing so . That value is, in 1/1024thas and ia the error that was recoded on internal oscillator frequency during factory cal. 
It rarely exceeds +/- 1 or 2. on the 2-series (it;'s 2-3x bigger for the 0/1-series largely because 2-series has 128 steps and 0/1 has only 64).
